File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        
        
        sort(nums.begin(), nums.end());
        int N = nums.size();
        
        int i=0, j=1;
        
        
        int ctr=0;
        while(j<N){
            
            //skip same elements
            while(j < (N-1) && nums[j]==nums[j+1]){
                j++;
            }
            
            int diff = abs(nums[j]- nums[i]);
            
            if(diff==k){
                ctr++;
                
                i++, j++;
            }else if(diff < k){
                j++;
            }else{
                i++;
            }
            
            if (i == j)
			    j++;
		
            
        }
        
        
        return ctr;
    }
};

File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.java
class Solution {
    public int findPairs(int[] nums, int k) {
        
        //https://leetcode.com/problems/k-diff-pairs-in-an-array/discuss/1756933/An-explanation-Going-from-O(NlogN)-greater-O(N)
        
        Map<Integer, Integer> map = new HashMap<>();
        
        for(int num : nums){
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        int count = 0;
        for(int x : map.keySet()){
            if(k > 0 && map.containsKey(x - k) || k == 0 && map.get(x) > 1) count++;
        }
        return count;
        
        
    }
}